### ISSUE TITLE
Support heterogeneous dissolution models to print.netdx and plot.netdx

### ISSUE THREAD

**Issue by sgoodreau** [EpiModel/EpiModel#508]
specifically nodematch and nodemix

**Comment by sgoodreau** [comment]
Howdy - just to make sure I don't create any confusion -- when I went to create a new branch a few weeks ago to work on this it seemed to me that you hadn't pulled in the branch for the previous work yet, after all the confusion around it.  You were out at the time so I held off mentioning it.  Now that I look again it seems as if relational_age2 is gone and I'm presuming merged in, and thus I can safely start a new branch off of master to do this work. Does that match your understanding?

**Comment by smjenness** [comment]
Yes, your earlier work was merged into master and you can create a new branch for this new work. Thanks!

**Comment by sgoodreau** [comment]
This is getting rapidly getting more interesting/complicated than I anticipated. 

First off is to confirm my understanding, as reflected in the name of this issue: the goal is to add these heterogeneous models for `netdx`, but not for `netsim`, since the latter are complicated by the fact that the attributes involved with the relevant model terms can be changing in `netsim`, but not in `netdx`. Yes?

If so, then the issue is that currently, both `print.netdx` and `print.netsim` call the helper function `make_dissolution_stats`.  This function does not currently include info about the network in its input arguments, because that was never needed before.  Now, for the `netdx` version, it is. My proposal is:

- add an argument to `make.dissolution.stats` that passes in a network
- when the function is called by `netdx`, pass in the static network associated with the simulation
- when the function is called by `netsim` pass in the dynamic network associated with the simulation
- if it's a homogeneous (edges-only) model, ignore the network, regardless of its form, and return homogeneous diss stats
- if it's a heterogeneous model, and the network is static, use that network to calculate the disagreggated diss stats
- if it's a heterogeneous model, and the network is dynamic, return the same error we currently have that calculations on heterogeneous are not supported in this case.

I think that's as logical and elegant as I can make it. Does that sounds good to you folks, @smjenness @AdrienLeGuillou, or can you think of better alternatives?

**Comment by sgoodreau** [comment]
Ooh, I have a slight tweak that will probably be better and a bit more elegant.  And without additional work it will allow the heterogeneous functionality to work with `netsim` as well, as long as the relevant attribute isn't a `TEA`.  Stay tuned.

**Comment by martinamorris** [comment]
Um, isn't age one of the vars often used for hetdiss?

**Comment by smjenness** [comment]
@sgoodreau : the above sounds like a good approach. Fine to support fixed attributes for now, perhaps supporting time-varying attributes in the future (although I don't know how because we don't routinely store TEAs for full-scale sims, but maybe that's a moot point because this is all for non-tergmLite simulations only anyway)...

**Comment by sgoodreau** [comment]
@martinamorris - I'm not actually sure anyone uses age for `hetdiss`, despite much discussion, given the many various theoretical and statistical issues we're still working out.  Note that `absdiff` isn't even an option in `EpiModel` for inclusion in a dissolution model, and most age-based models include that.  And even though 'nodefactor' *is* an option, it's not even clear in that case what diagnostics one will be interested in - we've come up against this a bunch but never fully laid it all out.

More broadly -- this ticket is for adding diagnostics for `netdx` specifically, since that is, for the most part, a piece of low-hanging fruit that's been sitting around for a very long time.  No attributes change during `netdx` at all. The only reason why attribute-changing is coming up here is because I realized that there are some places the print and plot helper functions for `netdx` call on lower-level functions that are shared with the same helper functions for `netsim`.  And in the process I realized that I could also add the heterogeneous output for `netsim` just as easily, in those cases where attributes also aren't changing.

Decision-points include:
- If we think it's confusing to show disaggregated diagnostics for `netsim` when the attributes are fixed vs when they're not, we can choose to just leave them out for `netsim` altogether for now.
- If we want to think about adding in netsim diagnostics for changing attributes, then we have to think those through very carefully, and it needs to build on our broader theoretical discussions about age and demography in these models.  Indeed, I would argue that we should move away from a focus on relational durations in such cases, and focus pretty much solely on the dissolution rate diagnostics -- as long as the models remain memoryless then this is relatively straight-forward.  Non-memoryless is yet another set of work beyond that. If this is a priority, however low, then I can file tickets on them. Thoughts?
- Either way, to wrap up this ticket, we will eventually need to have a discussion about what we should be showing for 'nodefactor' - a matrix of mean durations for every type?  Same for dissolution rates?  This is a lot of info, and I'm not sure if it will add more clarity or confusion for most users.

**Comment by sgoodreau** [comment]
On a recent call, we decided to deprecate and then remove the functionality for `nodefactor` in the dissolution models.  See issue #647.

**Comment by sgoodreau** [comment]
@AdrienLeGuillou and @smjenness 

Since I fell off the wagon working on this for way too long, I need to pull in updates from main. This is my first time resolving merge conflicts and it's fun.  Thankfully all but one are related to Karina's documentation, and are easy to integrate. 

The one that isn't is the code that Adrien wrote for issue #609, designed to pad out the mean ages vector when it wasn't long enough, in order to fix issue #605. Both of the functions relevant to that issue (`make_dissolution_coefs` and `edgelist_meanage`) change a lot as part of building out the functionality for the heterogenous dissolution diagnostics (e.g. among other things, the vector becomes a matrix).  I suspect that those won't obviate the problem that led to issue #605 and #609, and it will also need to be addressed in this new code with similar padding out. I could just follow the logic in Adrian's code blindly and extend it to the case of a matrix, but I'd rather understand the issue. You described it as arising "when the network is not simulated at the end of the model." What set of arguments or sequence of commands would have led to that happening?

**Comment by AdrienLeGuillou** [comment]
@sgoodreau this situation was specific to Greg project where he would stop the network simulation but continue to run the model for some time to get some follow up information. The state of the network at this point became irrelevant to the analysis and so re-simulating the network would slow down the process significantly.

To stop the network we used the "updater" system we showcased the other day to set `control$resimulate.network` to false at a given time step. ([see the relevant vignette section here])(https://github.com/EpiModel/EpiModel/blob/main/vignettes/model-parameters.Rmd#L315-L463)

**Comment by AdrienLeGuillou** [comment]
What is the part that becomes a matrix? PR#609 should not change the structure of any objects (unless I made a mistake of course)

**Comment by sgoodreau** [comment]
Thanks @AdrienLeGuillou.  To be clear - nothing in your code turns the vector into a matrix.  I meant that the changes I am going to make in this section of code as part of *this* issue (#508), will turn it into a matrix, at least in those cases where the dissoultion formula contanis more than one term.  That is, because we will be calculating mean durations for multiple subtypes of dyads at each timestep. So I will need to update your "Greg-specific" padding out to the matrix case too jut to be safe.

But this does raise an important question: with this functionality added, what should happen in the case of the simple homogeneous dissolution?  Should it return a vector (for backwards compatibility), even though this means that the output will sometimes be a vector and sometimes a matrix, with all of the issues that will create?  Or should it be changed to a 1 x t matrix, breaking some people's code but being internally consistent?  I favor the latter, but it's worth discussing the trade-offs and making thechoice thoughtfully.  @smjenness? @martinamorris? @gknowlt?

**Comment by AdrienLeGuillou** [comment]
I definitively favor the latter as well. Debugging with a function that returns sometimes a vector and sometimes a matrix is really hard

**Comment by sgoodreau** [comment]
@AdrienLeGuillou, @smjenness 

Out of curiosity, I notice the following conditional in `print.r` (also in `plot.r`):
```
    if (x$coef.diss$dissolution == ~ offset(edges)) {
      print_nwstats_table(x$stats.table.dissolution, digits)
      if (x$coef.diss$model.type == "hetero") {
        cat("----------------------- \n")
        cat("* Heterogeneous dissolution model results averaged over")
      }
    } else {
      cat("Not available when:")
      cat("\n- dissolution formula is not `~ offset(edges)`")
      cat("\n")
    }
  }
```  

All of this is going away as I write the code for heterogeneous models, and I could just delete it and not worry about it. But I'm very curious, why were there two checks here: `if (x$coef.diss$dissolution == ~ offset(edges)` and `if (x$coef.diss$model.type == "hetero")`?  And under what circumstances could those both have been true?  I can't think of any. That is, I thought `model.type` was only `hetero` when there was a second term, as confirmed by the code below:

```
model.type <- NA
  if (form.length == 1 && t1.edges == TRUE) {
    model.type <- "homog"
  } else if (form.length == 2 && t1.edges == TRUE &&
             t2.term %in% c("nodematch", "nodefactor", "nodemix")) {
    model.type <- "hetero"
  } else {
    model.type <- "invalid"
  }
```

AFAUK, is it safe to just delete this whole thing as I add in the conditionals?

More generally, I think with this functionality being added, it may be time to retire the `model.type` variable altogether, unless you see other important uses for it.

**Comment by smjenness** [comment]
That must have been a typo to have both the dissolution model be edges and model type hetero. This was a block of code that @AdrienLeGuillou edited, so I will ask him to comment/review when he returns from vacation next week.

**Comment by AdrienLeGuillou** [comment]
@sgoodreau, @smjenness 

I looked at the PR where I added this double check, I think at the time I simply added the check for `x$coef.diss$model.type == "hetero"` and kept the existing code exactly as it was to avoid unexpected regressions.

> AFAYK, is it safe to just delete this whole thing as I add in the conditionals?

I agree, this conditional part was a temporary fix before the full conditional system was built

**Comment by sgoodreau** [comment]
Thanks.  OK next question, @smjenness @AdrienLeGuillou.  It relates to the plotting, which I'm working on now, for both duration and dissolution.  So many different things to consider, I'm amazed that you had the fortitude to write all these options in the first place.

Q: Do you want users to be able to specify a subset of statistics they want to plot (similar to formation), or no?  In the case of a `nodematch` term, it's really no big deal, since there are just going to be two outcomes (`nodematch T` and `F`).  But for `nodemix` one could end up with quite a lot of groups, on scales that are hard to read simultaneously.  (Although users could use `joined=FALSE` to solve this). I don't have a deep sense yet of how complicated it'll be. Probably not hugely, just lots of little details. For one thing, we don't have quite the standardized language for referring to the entries in the same way that we do the formation terms, although I imagine we could use numbers.

**Comment by sgoodreau** [comment]
@smjenness @AdrienLeGuillou Things are moving along here. Calculations and table printing all work, now it's just plotting, documenting, formalizing the testing. Hoping to wrap up next week.

Quick inter-related questions: 

(1) My memory says that we agreed to no longer export `edgelist_meanage`, since it has grown in complexity a bit as more things need to be passed into it, and is less reobust to random usage. But we will still give it a help page.  Does this match your memory?  (I think the discussion happened off github, one more reason to always have these discussiosn here).

(2) If the answer to 1 is yes: the net.utils file starts with the comment "Exported Functions".  Should I move edgelist_meanage to another file then? If so, which?

**Comment by smjenness** [comment]
> Q: Do you want users to be able to specify a subset of statistics they want to plot (similar to formation), or no? 

Probably not to subsetting. the stats option for the formation model has to do selecting whole ergm terms (e.g., a nodefactor term altogether, including its X number of attribute levels). I believe you are talking about subsetting levels within a term. Juice not worth the squeeze.

> My memory says that we agreed to no longer export edgelist_meanage, since it has grown in complexity a bit as more things need to be passed into it, and is less reobust to random usage. But we will still give it a help page. Does this match your memory? (I think the discussion happened off github, one more reason to always have these discussiosn here).

I believe that's correct. There is no end user use case for this function. And they can use `:::` if they really want it.

>  If the answer to 1 is yes: the net.utils file starts with the comment "Exported Functions". Should I move edgelist_meanage to another file then? If so, which?

Please just remove the code comment. Everything is self-explanatory with the roxygen tags now.

**Comment by sgoodreau** [comment]
@smjenness @AdrienLeGuillou @martinamorris @chad-klumb 

Would you say that EpiModel supports dissolution models with `nodematch(diff=T)` or not?  In my head the answer was no, as I hadn't ever seen any examples in NME or our tutorials or `test_that netdx` that included them, and such models can be handled with `nodemix`. So I wrote my additions to not include that case. Turns out the remaining `testthat` errors are from a `netest` check that includes such an example. I need to revise the code either way, but how depends on whether we want to support `nodematch(diff=T)`.  If we don't support it, then I just need to add in a check early on to return an error, and then revise the relevant test.  If we do support it, then I'll need to go back and add in this case throughout.  Should be just a matter of an afternoon of work.  Either fine with me, although it might take me a few days before I can get to it.  I lean towards including it.

**Comment by chad-klumb** [comment]
I would think it should be supported, especially if existing checks don't prevent its use.

**Comment by AdrienLeGuillou** [comment]
I don't think there is anything that prevents it. Theoretically, everything that can be done with the networks is supported as long as they give a similarly formatted  object at the end.

**Comment by sgoodreau** [comment]
Copied from Steve's email:

@smjenness @chad-klumb @AdrienLeGuillou 

OK, I added functionality for `nodematch(diff=T)`, and the tests that were failing before are now passing.

But -- I'm down to two tests failing, and they're failing because the tests need to be updated. But I can't quite understand the logic of what the test is doing and how it's doing it, so I don't want to monkey with it.

Specifically it's `test-netest.R` >> `test_that("environment handling in non-nested EDA")`. The point here is something complicated about environments, clearly. The issue is that the creation of `netdx_2` and `netdx_4` both fail now, but for a reason other than that which the test is expecting. They fail because the length of the `diss_coef` vector doesn't match what it should, now that `netdx` checks on that. But I can't just change the length of `diss_coef` here, because that inteferes with the whole point of the way it's currently passed and the environments drama that creates. So I'm stuck.

Sam or someone, can you take a look at this and fix the test, or give me advice on how to proceed?

Otherwise I do believe I am done at long last!

**Comment by sgoodreau** [comment]
Reply from @AdrienLeGuillou:

I checked the tests, apparently the calls are not supposed to throw errors, and they do not in the main branch.

We should actually replace

```
expect_error(netdx_2 <- netdx(netest_2, nsims = 2, nsteps = 5, dynamic = TRUE), NA)

```

with

```
netdx_2 <- netdx(netest_2, nsims = 2, nsteps = 5, dynamic = TRUE)
```

as the first one uses `except_error(..., NA)` to check that no error is occuring. It's quite confusing.

For the call itself, it appears that `pages` calculated in `netdx` is filled with NAs in this branch. Which is not the case on `main`.

This occurs in the `make_dissolution_stats` function (R/netdx.R, line 512) whith this call: `get_vertex_attribute(diag.sim[[x]], diss_attr_name)` that actually returns a vector of NA as there is no vertex attribute called "a", only "age", "race", "vertex.name". "a" is the value stored in `diss_attr_name`.

I hope this helps.

**Comment by sgoodreau** [comment]
Reply from @chad-klumb:

`a` is a variable name; its value needs to be looked up in the formula environment.

**Comment by sgoodreau** [comment]
Reply from @sgoodreau:

Thanks all, for your quick replies. And I'm not sure why I didn't send this on Github; I'll move this convo there after this email.

I'll admit that environments have never been my strong point in R, to say the least.

That is confusing that `expect_error(, NA)` means "do not expect an error", rather than "expect an error that contains NA in it."  I guess the latter would have quotes around NA.

In any case, the fact that pages contains NAs now is a feature, not a bug.  This test is passing a vector of dissolution coefs that is the wrong length given the number of attribute values.  It used to not matter because `print.dx` and `plot.dx` just averaged over all ties, but now they actually calculate for each sub-group, and for them to do that netdx itself does some checks and data processing.

The reason that the vector is the wrong length is that it is just borrowed from one attribute (age) to another (race) in the middle of the code.

So it needs to be fixed to have the right length in `dissolution_coefs`.  But it needs to do it in a way that it still tests the functionality about environments that Sam (?) wrote this to test.  And since I don't 100% get this part, I'm loath to just edit it myself.

With that info in hand, do any of you see the best way to revise the tests?

**Comment by sgoodreau** [comment]
Reply from @chad-klumb:

I wrote the tests in question.

Can you explain in what sense the dissolution coefficients are of the wrong length?  `coef_diss` (with a duration vector of length 4) is used with "age" (3 unique attribute values), and `coef_diss_2` (with a duration vector of length 6) is used with "race" (5 unique attribute values).

**Comment by sgoodreau** [comment]
Reply from @sgoodreau:

OK, now I'm more confused than ever!

I'm looking at line 278:

```
netest_2 <- netest(nw = nw,
                     formation = ~edges + nodematch(r, diff = TRUE),
                     target.stats = c(500, 50, 50, 90, 40, 60),
                     coef.diss = coef_diss,
                     nested.edapprox = FALSE,
                     from.new = "a")
```

So from what I can see, `coef.diss` is being used with `netest_2` (not `coef_diss_2`, which is defined later), and `coef.diss` at that point has four coefficients, and includes the attribute `a` for age (lines 259 and 265).  I thought maybe that was by design and had something to do with what you're testing, but I'm not familiar with the arguments in lines 282 and 283, so I'm not sure.  Maybe it's not by design, and you meant to use `coef_diss_2`, and it just didn't show up as a problem until I added the new checks in `netdx`?

netest_4 may be a different story, but we can figure that out after this one.

Transferring this chain over to github now for posterity; please reply there.

Thanks!
Steve

**Comment by sgoodreau** [comment]
@chad-klumb 

I just realize that we're probably talking past each other.  That is, you're working in a world where the dissolution model no longer needs to be a subset of the formation model.  I haven't made that mental switch, so am looking at the formation model and assuming that the dissolution model is expecting the same, hence why in my brain `netest_2` is a model that is supposed to have `nodematch("race")` in both formation and dissolution and it's mismatching.  Let me take another look.....

**Comment by chad-klumb** [comment]
The association between durational vector and vertex attribute is made at the time `dissolution_coefs` is called, here:

https://github.com/EpiModel/EpiModel/blob/53bd16bbc5232678f7fc1c1521e40220300b68b1/tests/testthat/test-netest.R#L257-L265

and here:

https://github.com/EpiModel/EpiModel/blob/53bd16bbc5232678f7fc1c1521e40220300b68b1/tests/testthat/test-netest.R#L287-L295

As you say, in the non-nested case, what attribute is being used in the dissolution model cannot be inferred from looking at the `formation` argument to `netest`.

**Comment by sgoodreau** [comment]
@chad-klumb @smjenness @AdrienLeGuillou 

OK, ignore everything I've said until now. That was me not realizing that the dissolution model could have a different attribute in its `nodematch` term than the formation model does. Mea culpa.

It seems like the actual issue comes from the fact that I wrote my code in a way that doesn't handle the environent issues that Chad is testing on.

Context: it used to be that `edgelist_meanage` just needed the edgelist passed to it, and that is how it works in `main`.  But in order for the function to now calculate the disaggregated durations given the heterogenous model, it also needs to have the dissolution model info and the attribute info passed in.  This is one of the things I added.  One can see this in the line in `netdx` in the `netdx_heterodiss` branch:

```
pages <- sapply(seq_along(sim.df), function(x) {
                      meanage <- edgelist_meanage(el=sim.df[[x]], diss_term=diss_term,
                                   diss_attr=if(is.null(diss_term)) NULL else  
                                     get_vertex_attribute(diag.sim[[x]], diss_attr_name),
                                   diss_arg=if(!exists("diss_arg")) NULL else diss_arg)
```

where `diss_attr` is the set of values for the attribute of interest (named `diss_attr_name`), which is NULL if the dissolution model is homogeneous, and `diss_arg` is a flag (TRUE/FALSE) for the diff type, which is NULL unless the model has a `nodematch` term in it.  This code is not the most elegant for sure, becuase (1) I am very far away from being an elegant R programmer, and (2) this and all the other changes I made are being fit into a system in which multiple functions needed less info than they now do, in ways that disrupt the existing architecture as little as possible given my abilities.

It seems that the way I've written this, and with the way Chad has set up the test, this code is now looking for an attribute named `a`, whereas it needs to be looking for an attribute named `age`, where `age` is a string stored in the variable named `a`.  Alas, this is where my comfort level as an R programmer quickly breaks down.

@chad-klumb Do you seen an easy way to fix this code to do what you need it to do here?

**Comment by chad-klumb** [comment]
Are term arguments being parsed as strings in this code?

https://github.com/EpiModel/EpiModel/blob/c515ddbbbf6de8cdfb1806554f137402532a87cd/R/netdx.R#L476-L506

In general, to correctly evaluate the term arguments (and thus to know what vertex attribute is being referred to), one needs to go through the `ergm_model` song and dance:

https://github.com/statnet/ergm/blob/master/R/ergm_model.R

**Comment by sgoodreau** [comment]
> Are term arguments being parsed as strings in this code?

> [code snipped]

They are.
 
> In general, to correctly evaluate the term arguments (and thus to know what vertex attribute is being referred to), one needs to go through the `ergm_model` song and dance:
> 
> https://github.com/statnet/ergm/blob/master/R/ergm_model.R

OK.  Note that I copied this code from two places (`dissolution_coefs` and `diss_check`), both of which were presumably written by @smjenness way back at the start of EpiModel's life. And then I added additional lines to it for the added functionality.  So this is probably a broader issue, that just has never come up before because none of us have thought to do the kind of environment checking that you are, and no users have used it that way either.  If there's a more robust way this all needs to be done, then I think it'll need to be tackled in all three places.

Note my TODO line just above that bit of code; I filed an issue (#661) to consider revising one or both of those earlier functions to save more info and pass it out, so that we don't need this redundant code here.  Sam agreed to the proposal.  So if it turns out we need to re-write this code now anyway before this branch can be merged, then maybe it's worth tackling that.

Not sure if the best strategy here is to do all of this before merging this branch in.  Probably, since I think the only other option is to remove the tests temporarily.

But I'm left wondering why the similar code in `dissolution_coefs` and `diss_check` doesn't trip any of your tests?  Is there a fundamental difference, or do thetests just not check on them in the same way?

In any case, this seems like a job that you're much better positioned to tackle than me, @chad-klumb.  @smjenness, do you agree?

Mentioning @martinamorris here since she hasn't seen any of this yet and may be interested.

**Comment by sgoodreau** [comment]
One additional thought: EpiModel is not a statnet pakage per se, and there are many ways that EpiModel does not support the general functionality present in the statnet pacakges, but is much more restritive in its expectations.  So one option here is to leave EpiModel as is, requiring the input to take a specific form, and thus simply remove the tests.  Again, environments are not my strong suit, so I don't know what the use cases are for the functionality that you're testing that doesn't work in EpiModel now, and if those are things that EpiModel users are likely to ever actually want to do.  

Everyone?

**Comment by chad-klumb** [comment]
Parsing arguments as strings would be an issue not just for e.g. `x <- "race"; ~nodefactor(x)` but also `~nodefactor(~race)` and various other uses of the nodal attributes and levels API.  I think we've all used that, and it's important for it to behave as expected in `EpiModel` (i.e., consistent with `statnet`).

If I needed to categorize toggles based on dyad type I would probably use `ergm.godfather`, which allows one to obtain the relevant change statistics (and thus dyad types) without getting into details of term implementations or argument parsing.  In this particular case, I would look to combine this work with #490, since both involve major changes to `netdx`, and it would be nice to do the work only once, with both ends in mind.

At a glance it appears that `dissolution_coefs` is concerned with the `offset` decorator and `edges`, `node*` term names, not the arguments to the terms.  `diss_check` may require additional assumptions to justify its implementation; in general, the "string representations" of two formulas being the same is neither necessary nor sufficient for the corresponding models to be the same (e.g., the environments could be different, or different variable names could be bound to the same value).

**Comment by sgoodreau** [comment]
Thanks much @chad-klumb. 

Looks like we've reached the point where we definitely need input from @smjenness to decide on a path along which to move forward.  And @martinamorris and @AdrienLeGuillou and all other's thoughts welcome too, of course.

If you're just tuning in, the summary is: the new heterogeneous dissolution code is effectively finished, but it trips two tests that @chad-klumb has added in the meanwhile.  After some confusion on my part about the cause (laid out in multiple comments above), the actual issue gets uncovered and explained starting with the comment above that begins "OK, ignore everything I've said until now." In short, it's about how EpiModel processes the `formation` and `dissolution` formulas to extract the term names, names of associated attributes, and now also additional arguments (e.g. `diff=TRUE`).  I copied and modified existing EpiModel code to do the additional versions of this needed in additional places, and added some new pieces to it, but it looks like all of this code, old and new, is in violation of the best practices developed within the `statnet` packages to allow for much more robust versions of argument passing.

Given this, I'm starting to favor:
(1) removing Chad's two tests temporarily, since they get into the realm of checking on string passing and environments that seems to be violated by a bunch of long-standing EpiModel functionality.
(2) pull in this branch so that the work is done and integrated and we don't have to worry about it getting out of sync anymore
(3) then having Chad proceed to fix the string parsing as identified with his tests across all of `EpiModel` functionality, along with all the other aspects of `netdx` we decide are worth bundling in together.

**Comment by martinamorris** [comment]
Sounds like the right plan to me.

**Comment by smjenness** [comment]
@sgoodreau Thanks for all this! I read from the top, and then got your nice summary at the bottom. I think I probably originally developed the simple string processing a way back just to handle the very limited cases in our NME and applied HIV work, and it worked well enough there but I realized it was a bit duct taped together. I agree with your proposal of 1 and 2 in the short term, and then having @chad-klumb looking into 3 in the longer-term. Thanks!

**Comment by sgoodreau** [comment]
Great.  I'll comment out the two tests now in the `netdx_heterodiss` branch, push it, confirm that the branch passes all tests, and then send the pull request. Happy happy.
